homeassistant/components/fibaro/__init__.py

```diff
@@ -357,39 +355,78 @@ def _read_devices(self):
                 pass
 
 
-def setup(hass: HomeAssistant, base_config: ConfigType) -> bool:
-    """Set up the Fibaro Component."""
+async def async_setup(hass: HomeAssistant, base_config: ConfigType) -> bool:
+    """Migrate configuration from configuration.yaml."""
+    if DOMAIN not in base_config:
+        return True
     gateways = base_config[DOMAIN][CONF_GATEWAYS]
-    hass.data[FIBARO_CONTROLLERS] = {}
+    if gateways is None:
+        return True
+
+    # check if already configured
+    domains_list = hass.config_entries.async_domains()
+    if DOMAIN in domains_list:
+        return True
+
+    for gateway in gateways:
+        # prepare new config based on configuration.yaml
+        conf = {}
+        conf[CONF_URL] = gateway[CONF_URL]
+        conf[CONF_USERNAME] = gateway[CONF_USERNAME]
+        conf[CONF_PASSWORD] = gateway[CONF_PASSWORD]
+        conf[CONF_IMPORT_PLUGINS] = gateway[CONF_PLUGINS]
+
+        # import into config flow based configuration
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
 
-    def stop_fibaro(event):
-        """Stop Fibaro Thread."""
-        _LOGGER.info("Shutting down Fibaro connection")
-        for controller in hass.data[FIBARO_CONTROLLERS].values():
-            controller.disable_state_handler()
+    return True
 
-    hass.data[FIBARO_DEVICES] = {}
+
+def _init_controller(data: dict[str, Any]) -> FibaroController:
+    """Validate the user input allows us to connect to fibaro."""
+    controller = FibaroController(data)
+    if controller.connect():
+        return controller
+
+    raise Exception("Connection to fibaro failed")
```

Comment:
Please don't raise generic exceptions

Reply:
Replaced by returning a tuple of [connected, controller] which makes more sense.